pkg/jobmanager/jobmanager.go

```diff
@@ -45,3 +55,69 @@ func (jm *JobManager) LookupJob(id uuid.UUID) (*Job, error) {
 	}
 	return j, nil
 }
+
+func (jm *JobManager) StopJob(id uuid.UUID) error {
+	j, err := jm.LookupJob(id)
+	if err != nil {
+		return err
+	}
+	j.mu.Lock() // NOTE: We don't defer unlock as we want to wait after unlock unpon success.
+	if j.status != pb.JobStatus_JOB_STATUS_RUNNING {
+		j.mu.Unlock()
+		return nil
+	}
+	if j.cmd.Process != nil {
+		// Send the KILL to the process group, not just the process to avoid orphans.
+		// NOTE: This is POSIX compliant.
+		if err := syscall.Kill(-j.cmd.Process.Pid, syscall.SIGKILL); err != nil {
+			j.mu.Unlock()
+			if errors.Is(err, syscall.ESRCH) {
+				// If the process died as we were about to stop it, nothing to do.
+				// Don't set the status as stopped as it exited on it's own.
+				// This is an unavoidable "race" as we don't control the child process,
+				// it can die after the lock and before the kill. Nothing to worry about though.
+				return nil
+			}
+			return fmt.Errorf("process kill %d: %w", j.cmd.Process.Pid, err)
+		}
+	}
+	j.status = pb.JobStatus_JOB_STATUS_STOPPED
+	j.mu.Unlock()
+	<-j.waitChan
+	return nil
+}
+
+func (jm *JobManager) StreamLogs(ctx context.Context, id uuid.UUID) (io.Reader, error) {
+	j, err := jm.LookupJob(id)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a pipe for the caller to consume.
+	r, w := io.Pipe()
+
+	// Pause the broadcast, make a copy of the current output and subscribe the live logs.
+	//
+	// NOTE: Not ideal as it pauses the broadcast for every clients connected to that stream.
+	// In production, should consider a more advanced setup where we only pause the historical
+	// feed without pausing the stream of other clients.
+	// As it only pauses while making a copy of the output buffer and adding one entry to a map,
+	// it is acceptable for now.
+	j.broadcaster.Pause()
+	output := j.output.String()
```

Comment:
Could we avoid taking this copy?

Reply:
I don't believe so. We need the historical output in the client logs. If we didn't have a copy, we'd have a race as the broadcaster would keep writing to it.

If we were to keep the broadcaster paused, then we may avoid the copy.

In a production setting, this wouldn't be a concern though as we'd use a file on disc.

While not ideal, for the sake of the exercise, it does the job well.

Reply:
FWIW switching to a `strings.Builder` is a +3-4 diff and avoids copying the whole output on every `StreamLogs` call.

Reply:
arg, indeed, forgot that was a thing. updated.